### PR TITLE
partition: add directory sync to the save uboot.env file code

### PR DESCRIPTION
**Pull request by mvo5**
This branch is doing two things:
a) move the ubootenv handling code from my github branch into snapd itself, the code started as an experiment but now that we use it in production we really want it to be part of the snapd code to benefit from the extra reviews, extra tests etc
b) add a dir.Sync() when a uboot.env file is written

### CODE REVIEW

File: partition/ubootenv/env.go [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+	contentWithHeader, err := ioutil.ReadAll(f)
+	if err != nil {
+		return nil, err
+	}
+	crc := readUint32(contentWithHeader)
+
+	payload := contentWithHeader[headerSize:]
+	actualCRC := crc32.ChecksumIEEE(payload)
+	if crc != actualCRC {
+		return nil, fmt.Errorf("bad CRC: %v != %v", crc, actualCRC)
```
Comment (pedronis) [inline partition/ubootenv/env.go]: this path doesn't seem tested

File: partition/ubootenv/env.go [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+			continue
+		}
+		l := strings.SplitN(string(envStr), "=", 2)
+		if len(l) != 2 {
+			if flags&OpenBestEffort == OpenBestEffort {
+				continue
+			}
+			return nil, fmt.Errorf("cannot parse line %q as key=value pair", envStr)
+		}
+		key := l[0]
```
Comment (pedronis) [inline partition/ubootenv/env.go]: shouldn't we check that key is not empty?  especially given  we panic in iterEnv in such a case?

File: partition/ubootenv/env.go [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+func (env *Env) iterEnv(f func(key, value string)) {
+	keys := make([]string, 0, len(env.data))
+	for k := range env.data {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	for _, k := range keys {
+		if k == "" {
+			panic("iterEnv iterating over a empty key")
```
Comment (pedronis) [inline partition/ubootenv/env.go]: this is ok as long as make sure keys in data must never be empty by construction

File: partition/ubootenv/env.go (hunk 1 above)
Comment (mvo5) [inline partition/ubootenv/env.go]: Indeed, thank you! I added a test for this now.

File: partition/ubootenv/env.go (hunk 2 above)
Comment (mvo5) [inline partition/ubootenv/env.go]: Nice catch, thank you. I added code to error now when this line is empty.

File: packaging/ubuntu-14.04/not-installed [hunk 4]
```diff
@@ -1 +0,0 @@
-debian/tmp/usr/bin/uboot-go
```
Comment (zyga) [inline packaging/ubuntu-14.04/not-installed]: Nice :-)

File: partition/uboot_test.go [hunk 5]
```diff
@@ -23,16 +23,16 @@ import (
 	"os"
 	"time"
 
-	"github.com/mvo5/uboot-go/uenv"
```
Comment (zyga) [inline partition/uboot_test.go]: Do we need to update any vendor files?

File: vendor/vendor.json [hunk 6]
```diff
@@ -43,18 +43,6 @@
 			"revisionTime": "2015-02-12T09:37:50Z"
 		},
 		{
-			"checksumSHA1": "c1zgLuM2b+2vguxwz+XVMXPnXOY=",
```
Comment (zyga) [inline vendor/vendor.json]: Ah, I see now! (I looked at a partial diff before)

**Pull request merged**